File: demo-host/app/types/AiAssistantClientSpace.ts
import type { AssistantWidgetLang } from '~/types/AiAssistant'

/** What a visitor asked for, as the API types a request. */
export type AiAssistantClientRequestType = 'question' | 'quote' | 'appointment' | 'urgent' | 'other'

/** Where the business is with a request. */
export type AiAssistantClientRequestStatus = 'new' | 'handled' | 'dropped'

/**
 * One request as its client sees it; `received_label` is already in business time (« 14/09 à 10:05 »),
 * `appointment_slots` are the wished half-days of an appointment request (« lun. 28/09, matin »), to confirm,
 * and `appointment_booked` the appointment already in the agenda (« mar. 29/09 à 14:30 (Révision) »).
 */
export type AiAssistantClientRequest = {
  id: number
  type: AiAssistantClientRequestType
  status: AiAssistantClientRequestStatus
  name: string
  contact: string
  summary: string | null
  received_label: string
  received_outside_hours: boolean | null
  photo_urls: string[]
  appointment_slots: string[]
  appointment_booked: string | null
}

/** The latest monthly report of the assistant; its sentences come ready-made, like in the report email. */
export type AiAssistantClientReport = {
  month_label: string
  conversations: number
  requests: number
  quotes: number
  appointments: number
  urgent: number
  photo_requests: number
  outside_hours_pct: number | null
  languages_line: string | null
  handling_line: string | null
  top_questions: string[]
}

/** Where the client's subscription stands. */
export type AiAssistantClientSubscriptionStatus = 'incomplete' | 'active' | 'past_due' | 'canceled'

/**
 * The client's subscription; `cancel_scheduled` means paid until the period end but not renewed, and
 * `can_manage` tells whether the Stripe billing portal can open.
 */
export type AiAssistantClientSubscription = {
  status: AiAssistantClientSubscriptionStatus
  price_label: string
  period_end_label: string | null
  cancel_scheduled: boolean
  can_manage: boolean
}

/** The settings a client may change, defaults applied. */
export type AiAssistantClientSettings = {
  assistant_name: string
  languages: AssistantWidgetLang[]
  alert_phone: string | null
  alert_sms_enabled: boolean
  alert_email_enabled: boolean
  /** The request types texted at once; the others go by email only. */
  alert_sms_types: AiAssistantClientRequestType[]
  /** The window during which the SMS are held (Paris hours; equal hours = never held). */
  alert_quiet_start_hour: number
  alert_quiet_end_hour: number
}

/** A language the widget can speak, with its French name. */
export type AiAssistantClientLanguageOption = {
  code: AssistantWidgetLang
  label: string
}

/** Where the agenda stands: Google not configured on the server, not connected, connected, or its access lost. */
export type AiAssistantClientCalendarStatus = 'unavailable' | 'disconnected' | 'connected' | 'error'

/**
 * The agenda section: its connection, the booking settings (defaults applied), the choices offered, and the last
 * problem met with the agenda (`last_error`, cleared by a booking).
 */
export type AiAssistantClientCalendar = {
  status: AiAssistantClientCalendarStatus
  account_email: string | null
  calendar_id: string
  duration_minutes: number
  min_notice_hours: number
  appointment_types: string[]
  last_error: string | null
  duration_choices: number[]
  min_notice_choices: number[]
}

/** A client's booking settings edit; every field is optional. */
export type AiAssistantClientCalendarUpdate = Partial<
  Pick<AiAssistantClientCalendar, 'calendar_id' | 'duration_minutes' | 'min_notice_hours' | 'appointment_types'>
>

/** The Google consent page to open in a new tab. */
export type AiAssistantClientCalendarConnect = {
  url: string
}

/** An upcoming appointment the assistant booked; `start_label` is in business time (« mar. 29/09 à 14:30 »). */
export type AiAssistantClientAppointment = {
  id: number
  start_label: string
  type_label: string | null
  name: string
  contact: string
}

/** Everything the client-space page shows, as served by the API. */
export type AiAssistantClientSpace = {
  business_name: string
  assistant_name: string
  accent_color: string | null
  link_expires_label: string
  pending_count: number
  requests: AiAssistantClientRequest[]
  report: AiAssistantClientReport | null
  /** The example space a prospect opens from its demo page: fictional data, nothing to save. */
  is_example: boolean
  settings: AiAssistantClientSettings
  language_options: AiAssistantClientLanguageOption[]
  subscription: AiAssistantClientSubscription | null
  calendar: AiAssistantClientCalendar
  appointments: AiAssistantClientAppointment[]
  faq: AiAssistantClientFaqEntry[]
  unanswered: AiAssistantClientUnansweredEntry[]
}

/** A question the receptionist could not answer, with how often visitors asked it (dates in naive UTC). */
export type AiAssistantClientUnansweredEntry = {
  question: string
  count: number
  first_seen: string | null
  last_seen: string | null
}

/** An answer the business gave, which the receptionist now uses as is. */
export type AiAssistantClientFaqEntry = {
  question: string
  answer: string
  created_at: string | null
}

/** The two lists, as the API returns them after a change. */
export type AiAssistantClientFaqResponse = {
  faq: AiAssistantClientFaqEntry[]
  unanswered: AiAssistantClientUnansweredEntry[]
}

/** A client's settings edit; every field is optional. */
export type AiAssistantClientSettingsUpdate = Partial<AiAssistantClientSettings>

/** Where the page stands: loading, the space, an expired link (a new one can be emailed), an unknown link, or an outage. */
export type AiAssistantClientSpaceState = 'loading' | 'ready' | 'expired' | 'invalid' | 'unavailable'

/** The page's load result: its state, and the space when it opened. */
export type AiAssistantClientSpaceLoad = {
  state: AiAssistantClientSpaceState
  space: AiAssistantClientSpace | null
}

/** The Stripe billing portal session to open. */
export type AiAssistantClientPortal = {
  url: string
}

/** Whether a fresh link was emailed to the business. */
export type AiAssistantClientRenew = {
  sent: boolean
}

/** Where the « recevoir un nouveau lien » request of an expired link stands. */
export type AiAssistantClientRenewState = 'idle' | 'sending' | 'sent' | 'failed'

File: api/schemas/ai_assistant_client_space.py
"""Contracts of a sold assistant's client space (the magic-link page) and of its link, issued from the dashboard."""

from datetime import datetime
from typing import Annotated

from pydantic import BaseModel, Field, StringConstraints

from enums.ai_assistant_request import AiAssistantRequestStatus, AiAssistantRequestType
from enums.assistant_calendar_status import AssistantCalendarConnection
from enums.assistant_subscription_status import AssistantSubscriptionStatus
from enums.assistant_widget_language import AssistantWidgetLanguage
from schemas.ai_assistant_faq import AiAssistantFaqEntry, AiAssistantUnansweredEntry


class AiAssistantClientRequestItem(BaseModel):
    """One request as its client sees it."""

    id: int
    type: AiAssistantRequestType
    status: AiAssistantRequestStatus
    name: str
    contact: str
    summary: str | None = None
    # « 14/09 à 10:05 », business time (Paris).
    received_label: str
    received_outside_hours: bool | None = None
    photo_urls: list[str] = Field(default_factory=list)
    # Wished half-days of an appointment request (« lun. 28/09, matin »).
    appointment_slots: list[str] = Field(default_factory=list)
    # The appointment booked in the agenda (« mar. 29/09 à 14:30 (Révision) »).
    appointment_booked: str | None = None


class AiAssistantClientReport(BaseModel):
    """The latest monthly report of the assistant."""

    month_label: str
    conversations: int
    requests: int
    quotes: int
    appointments: int
    urgent: int
    photo_requests: int
    outside_hours_pct: int | None = None
    # The report email's own sentences (« français 72 %, anglais 20 % », « 31 demandes marquées traitées… »).
    languages_line: str | None = None
    handling_line: str | None = None
    top_questions: list[str] = Field(default_factory=list)


class AiAssistantClientSubscription(BaseModel):
    """The client's subscription, read from its local copy."""

    status: AssistantSubscriptionStatus
    # « 79 €/mois » or « 790 €/an », the price locked at subscription.
    price_label: str
    # « 12/10/2026 »: end of the paid period, when Stripe sent it.
    period_end_label: str | None = None
    # The client scheduled the end in the portal: paid until ``period_end_label``, not renewed.
    cancel_scheduled: bool = False
    # The Stripe billing portal can open (a Stripe customer is known).
    can_manage: bool


class AiAssistantClientSettings(BaseModel):
    """The settings a client may change, defaults applied."""

    assistant_name: str
    languages: list[AssistantWidgetLanguage]
    alert_phone: str | None = None
    alert_sms_enabled: bool
    alert_email_enabled: bool
    # The request types texted at once (the others go by email only), and the window during which the SMS are
    # held (Paris hours; equal hours = never held).
    alert_sms_types: list[AiAssistantRequestType]
    alert_quiet_start_hour: int
    alert_quiet_end_hour: int


class AiAssistantClientLanguageOption(BaseModel):
    """A language the widget can speak, with its French name."""

    code: AssistantWidgetLanguage
    label: str


class AiAssistantClientCalendar(BaseModel):
    """The agenda section: its connection and the booking settings (defaults applied)."""

    status: AssistantCalendarConnection
    account_email: str | None = None
    calendar_id: str
    duration_minutes: int
    min_notice_hours: int
    appointment_types: list[str] = Field(default_factory=list)
    # The last problem met with the agenda (« 24/09 à 10:05 : agenda introuvable… »), cleared by a booking.
    last_error: str | None = None
    duration_choices: list[int] = Field(default_factory=list)
    min_notice_choices: list[int] = Field(default_factory=list)


class AiAssistantClientAppointmentItem(BaseModel):
    """An upcoming appointment the assistant booked."""

    id: int
    # « mar. 29/09 à 14:30 », business time.
    start_label: str
    type_label: str | None = None
    name: str
    contact: str


class AiAssistantClientSpaceResponse(BaseModel):
    """Everything the client-space page shows."""

    business_name: str
    assistant_name: str
    accent_color: str | None = None
    # « 24/10/2026 »: the last day this link opens the space.
    link_expires_label: str
    pending_count: int
    requests: list[AiAssistantClientRequestItem] = Field(default_factory=list)
    report: AiAssistantClientReport | None = None
    settings: AiAssistantClientSettings
    language_options: list[AiAssistantClientLanguageOption] = Field(default_factory=list)
    subscription: AiAssistantClientSubscription | None = None
    calendar: AiAssistantClientCalendar
    appointments: list[AiAssistantClientAppointmentItem] = Field(default_factory=list)
    # The answers the business wrote, and the questions its assistant could not answer.
    faq: list[AiAssistantFaqEntry] = Field(default_factory=list)
    unanswered: list[AiAssistantUnansweredEntry] = Field(default_factory=list)


class AiAssistantClientSettingsUpdate(BaseModel):
    """A client's settings edit (partial; the alert mobile is read like in the dashboard)."""

    assistant_name: str | None = Field(default=None, min_length=1, max_length=64)
    languages: list[AssistantWidgetLanguage] | None = Field(
        default=None, min_length=1, max_length=len(AssistantWidgetLanguage)
    )
    # As typed; empty clears it (no SMS).
    alert_phone: str | None = Field(default=None, max_length=32)
    alert_sms_enabled: bool | None = None
    alert_email_enabled: bool | None = None
    alert_sms_types: list[AiAssistantRequestType] | None = None
    alert_quiet_start_hour: int | None = Field(default=None, ge=0, le=23)
    alert_quiet_end_hour: int | None = Field(default=None, ge=0, le=23)


class AiAssistantClientCalendarUpdate(BaseModel):
    """A client's booking settings edit (partial; choices checked by the service)."""

    calendar_id: str | None = Field(default=None, max_length=255)
    duration_minutes: int | None = None
    min_notice_hours: int | None = None
    # Longer kinds are cut to 40 characters by the service.
    appointment_types: list[Annotated[str, StringConstraints(max_length=200)]] | None = Field(
        default=None, max_length=6
    )


class AiAssistantClientCalendarConnect(BaseModel):
    """The Google consent page to open in a new tab."""

    url: str


class AiAssistantClientPortalResponse(BaseModel):
    """The Stripe billing portal session to redirect the client to."""

    url: str


class AiAssistantClientRenewResponse(BaseModel):
    """Whether a fresh link was emailed to the business (its address is never disclosed)."""

    sent: bool


class AiAssistantClientLinkRequest(BaseModel):
    """The operator issues a client-space link, and may email it to the business."""

    send: bool = False


class AiAssistantClientLinkResponse(BaseModel):
    """A fresh client-space link, and where it was emailed."""

    url: str
    expires_at: datetime
    # The business address it went to; None when not asked, or when it could not leave.
    sent_to: str | None = None
    send_error: str | None = None
